ERCS/erc-7841.md

```diff
@@ -0,0 +1,550 @@
+- In ***synchronous*** messaging protocols, chains have synchronized blocks (e.g., a block for each chain is produced every *t* seconds) and messages are received on the destination chain within the same block timeslot as they were sent from the source chain. In particular, a chain may send a message and read a response in one transaction within a block on a chain.
+- In ***asynchronous*** messaging protocols the restrictions of synchronous protocols do not apply. In particular, the chains sending and receiving messages may not have synchronized block timeslots and there may be a delay (measured in elapsed blocks) between the transaction on the source chain that sends a message and the transaction on the destination chain in which it is received.
+
+There is a wide range of ways in which both synchronous and asynchronous protocols may operate. For example, some protocols (particularly for asynchronous messaging) may require the source chain's block in which the message was sent to be finalized (e.g., settled on Ethereum) before it can be included within a block produced for the destination chain. Other protocols may allow for messages to be optimistically included in blocks with consistency checks delayed until settlement time (reminiscent of speculative execution).
+
+### Message Format
+
+All messages MUST follow this format:
+
+```solidity
+/// @title Metadata type
+/// @notice Metadata for a cross-chain message
+struct Metadata {
+    /// @notice The chain identifier of the source chain
+    uint32 srcChainId;
+    /// @notice The chain identifier of the destination chain
+    uint32 destChainId;
+    /// @notice The address of the sending party
+    /// @dev 32 bytes are used to encode the address. In the case of an
+    ///     Ethereum address, the last 12 bytes can be padded with zeros
+    bytes32 srcAddress;
+    /// @notice The address of the recipient
+    /// @dev 32 bytes are used to encode the address. In the case of an 
+    ///     Ethereum address, the last 12 bytes can be padded with zeros
+    bytes32 destAddress;
+    /// @notice The identifier for a cross-chain interaction session
+    /// @dev SHOULD be unique for every new cross-chain calls
+    uint128 sessionId;
+    /// @notice The message counter within an interaction session 
+    /// @dev SHOULD be unique within a session
+    /// @dev OPTIONAL for most asynchronous bridges where every message has a 
+    ///     distinct sessionId, simply set to 0 if unused
+    /// @dev E.g. In a cross-chain call: ChainA.func1 -m1-> ChainB.func2 -m2-> 
+    ///     ChainC.func3 -m3-> ChainB.func4, the subscript i in m_i is the nonce
+    uint128 nonce;
+}
+
+/// @title Message type
+/// @notice A cross-chain message
+struct Message {
+    /// @notice The message metadata 
+    Metadata metadata;
+    /// @notice Message payload encoded using RLP serialization
```

Comment:
It seems that `payload` can use any encoding, not just RLP. (The two examples both use ABI encoding.)

Reply:
Yes, thanks for catching that. we should remove the mentioning of RLP here.
Theoreticall yes you can use any encoding, but in practice on EVM chain, most likely just ABI encoding.